### PR TITLE
Update attention

### PR DESCRIPTION
**Pull request by dkobylianskii**
- Add `self_attn` option, should speed up things a bit
- Use functions from `flash_attn` package to unpad/pad sequences
- Switch to single weight/bias chunk
- Add option to switch backends, should be useful for future exports/non-gpu launch

All tests are passed

@samvanstroud @trahxam

### CODE REVIEW

File: src/hepattn/models/attention.py [hunk 1]
```diff
@@ -78,30 +130,26 @@ def __init__(
...
         self.dim = dim
+        self.bias = bias
         self.num_heads = num_heads
+        self.head_dim = dim // num_heads
         self.attn_type = attn_type
-        self.attn = ATTN_TYPES[attn_type]
         self.window_size = None
         self.value_residual = value_residual
         self.qkv_norm = qkv_norm
+        self.self_attn = self_attn
```
[the remark this answers is not in the document]
Comment (dkobylianskii) [reply src/hepattn/models/attention.py:148]: Sounds good

File: tests/models/test_posenc.py [hunk 2]
```diff
@@ -4,7 +4,8 @@
 import torch
 
 from hepattn.models.posenc import PositionEncoder, pos_enc, pos_enc_symmetric
-from hepattn.models.posenc_random import PositionEncoderRandom
+# TODO: Uncomment when implemented
+# from hepattn.models.posenc_random import PositionEncoderRandom
```
[the remark this answers is not in the document]
Comment (dkobylianskii) [reply tests/models/test_posenc.py]: Done

File: src/hepattn/models/attention.py [hunk 3]
```diff
@@ -67,6 +70,55 @@ def merge_masks(
     return merged_mask
 
 
+def projection_packed(
```
[the remark this answers is not in the document]
Comment (dkobylianskii) [reply src/hepattn/models/attention.py:73]: It works, but it breaks `flex` attention with dynamic compile due to some issue with torch
So we can use it with `dynamic=False` only

File: src/hepattn/models/attention.py [hunk 4]
```diff
@@ -122,11 +195,84 @@ def recombine_heads(self, x: Tensor) -> Tensor:
...
+
+        # Residual connection with initial values
+        if self.value_residual:
+            if not initial_values:
+                initial_values["v"] = v
+            else:
+                v = v * mix + initial_values["v"] * (1.0 - mix)
+        return q, k, v
+
+    def _varlen_attention(
```
[the remark this answers is not in the document]
Comment (dkobylianskii) [reply src/hepattn/models/attention.py:235]: Done

File: src/hepattn/models/attention.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (samvanstroud) [reply src/hepattn/models/attention.py:73]: Sounds good. Flex + nested needs some custom masking logic I think

File: src/hepattn/models/decoder.py [hunk 5]
```diff
@@ -62,7 +62,18 @@ def forward(self, q: Tensor, kv: Tensor, attn_mask: Tensor | None = None, q_mask
...
             kv = self.kv_dense(kv)
 
         return q, kv
+
+    def set_backend(self, attn_type: str) -> None:
+        """
+        Set the backend for the attention layers.
+        This is useful for switching between different attention implementations.
+        """
+        self.q_ca.fn.set_backend(attn_type)
```
[the remark this answers is not in the document]
Comment (dkobylianskii) [reply src/hepattn/models/decoder.py:75]: We wrap attention in `Residual` and it's stored in `fn` field
